internal/pkg/service/stream/storage/level/local/writer/writer.go

```diff
@@ -1,42 +1,360 @@
+	compressedMeter   *size.MeterWithBackup
+	uncompressedMeter *size.MeterWithBackup
+}
+
+func New(
+	ctx context.Context,
+	logger log.Logger,
+	clk clock.Clock,
+	cfg Config,
+	slice *model.Slice,
+	file writechain.File,
+	dirPath string,
+	filePath string,
+	syncerFactory disksync.SyncerFactory,
+	formatWriterFactory FormatWriterFactory,
+	volumeEvents *Events,
+) (out Writer, err error) {
+	w := &writer{
+		dirPath:  dirPath,
+		filePath: filePath,
+		logger:   logger.WithComponent("slice-writer"),
+		slice:    slice,
+		events:   volumeEvents.Clone(), // clone volume events, to attach additional writer specific events
+		closed:   make(chan struct{}),
+		writeWg:  &sync.WaitGroup{},
+	}
+
+	// Close resources on error
+	defer func() {
+		if err != nil {
+			if w.syncer != nil {
+				_ = w.syncer.Stop(ctx)
+			}
+			if w.chain != nil {
+				_ = w.chain.Close(ctx)
+			}
+			if w.completedWrites != nil {
+				_ = w.completedWrites.Close()
+			}
+		}
+	}()
+
+	// Create counters
+	{
+		// In progress writes counter
+		w.acceptedWrites = count.NewCounter()
+
+		// Successful writes counter, the value backup is periodically saved to disk.
+		// In the case of non-graceful node shutdown, the initial state is loaded from the disk after the node is restarted.
+		// In that case, the statistics may not be accurate, but this should not happen, and we prefer throughput over the atomicity of statistics.
+		w.completedWrites, err = count.NewCounterWithBackupFile(ctx, clk, logger, filepath.Join(dirPath, CompletedWritesCounter), cfg.Statistics.DiskSyncInterval)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Create empty chain of writers to the file
+	{
+		w.chain = writechain.New(logger, file)
+	}
+
+	// Add a buffer before the file
+	{
+		if cfg.FileBuffer > 0 {
+			w.chain.PrependWriter(func(w io.Writer) io.Writer {
+				return limitbuffer.New(w, int(cfg.FileBuffer.Bytes()))
+			})
+		}
+	}
+
+	// Measure size of compressed data
+	{
+		backupPath := filepath.Join(dirPath, CompressedSizeFile)
+		_, err = w.chain.PrependWriterOrErr(func(writer io.Writer) (out io.Writer, err error) {
+			w.compressedMeter, err = size.NewMeterWithBackupFile(ctx, clk, logger, writer, backupPath, cfg.Statistics.DiskSyncInterval)
+			return w.compressedMeter, err
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Add compression if enabled
+	{
+		if compConfig := slice.LocalStorage.Compression; compConfig.Type != compression.TypeNone {
+			// Add compression writer
+			_, err = w.chain.PrependWriterOrErr(func(writer io.Writer) (io.Writer, error) {
+				return compressionWriter.New(writer, compConfig)
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			// Add a buffer before compression writer, if it is not included in the writer itself
+			if cfg.InputBuffer > 0 && !compConfig.HasWriterInputBuffer() {
+				w.chain.PrependWriter(func(w io.Writer) io.Writer {
+					return limitbuffer.New(w, int(cfg.InputBuffer.Bytes()))
+				})
+			}
```

Comment:
Important buffer before the compression, BUT! only, if there is no buffer inside the compression writer.

CSV rows can be pre-generated to the buffer before the compression writer.